### DISCUSSION TITLE
string/immutable

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#216] [Announcements]
# string/immutable

本文深入探讨了Java String 类的不可变性及其背后的设计原则。我们将了解不可变字符串如何提高代码的安全性、性能和可维护性，以及为什么Java选择了这种设计。探索String类的内部实现，理解为何Java字符串是不可变的。

https://tobebetterjavaer.com/string/immutable.html

**Comment by asdfs233** [comment]
在证明string不可变的安全性的那个示例，有证明什么吗    哪怕username，password都不是String ，是其他类型，直接改credentials[0]  credentials[1] 的值， 都不会影响原user对象吧。。       有没有大佬教一下

**Reply by Eivae** [reply]
我也是有点疑惑

**Reply by ZJRookie** [reply]
我感觉示例确实有问题，示例中的操作只是将另一个字符串的引用赋值给了credentials[0]，并未对原值进行修改，字符串不可变主要原因我觉得是没有提供可以操作底层char[]数组的api。

**Reply by ustcjsj** [reply]
String是引用数据类型，存储的是地址，直接赋值就是给它一个新地址，所以这个例子是挺搞笑的

**Reply by Git-jimmyLee-hub** [reply]
他应该是想说，如果string不是特殊的引用类型，像数组一样。例如每个用户密码用
String[] password = new String[1];存储，假设原先的password[0] = "abc";
当外界调用：
String[] temp = password;
temp[0] = "def";
那么password就会改变成”def“，密码被篡改。

**Comment by DevDreamer-Code-Ye** [comment]
我觉得博主例子没问题啊，如果向你说的那样换成别的数据类型，如果是基本数据类型值传递那肯定不会改变原值，但如果是自定义的引用类型通过址传递，那肯定会改动原始值的，所以话再说回来，String作为引用类型，博主的例子恰好说明了他的不可变性，因为以上的例子，字符串是存放在字符串常量池中，当尝试修改原始值时，则又在池中创建了新的字符串对象，并没有覆盖，所以原始值指向的还是原来的地址，这恰好说明了不可变性。
个人见解，不一定对，如果博主看见了的话，希望可以帮忙改正一下，谢谢！

**Reply by maozibuer** [reply]
问题就是作者相当于直接给字符串数组赋值，跟字符串没有什么关系，要想展示字符串的不可变性感觉用concat更好，个人理解，水平有限

**Comment by umico8832** [comment]
那个例子仅仅演示了 Java 中最基本的引用赋值，和不可变性几乎没有关系

**Comment by 3524315495** [comment]
误人子弟,  修改用户名密码案例主要说明的是"引用独立性"，而不是"字符串不可变性"
